pycoast/cw_base.py

```diff
@@ -881,17 +888,60 @@ def add_overlay_from_dict(self, overlays, area_def, cache_epoch=None, background
                           outline=outline, minor_outline=minor_outline,
                           minor_is_tick=minor_is_tick,
                           lon_placement=lon_placement,
-                          lat_placement=lat_placement)
+                          lat_placement=lat_placement,
+                          **grid_kwargs)
 
         if cache_file is not None:
-            try:
-                foreground.save(cache_file)
-            except IOError as e:
-                logger.error("Can't save cache: %s", str(e))
-            if background is not None:
-                background.paste(foreground, mask=foreground.split()[-1])
+            self._write_and_apply_new_cached_image(cache_file, foreground, background)
         return foreground
 
+    @staticmethod
+    def _apply_cached_image(cache_file, cache_epoch, background, regenerate=False):
+        try:
+            config_time = cache_epoch or 0
+            cache_time = os.path.getmtime(cache_file)
+            # Cache file will be used only if it's newer than config file
+            if config_time is not None and config_time < cache_time and not regenerate:
+                foreground = Image.open(cache_file)
+                logger.info('Using image in cache %s', cache_file)
+                if background is not None:
+                    background.paste(foreground, mask=foreground.split()[-1])
+                return foreground
+            logger.info("Regenerating cache file.")
+        except OSError:
+            logger.info("No overlay image found, new overlay image will be saved in cache.")
+        return None
+
+    @staticmethod
+    def _write_and_apply_new_cached_image(cache_file, foreground, background):
+        try:
+            foreground.save(cache_file)
+        except IOError as e:
+            logger.error("Can't save cache: %s", str(e))
+        if background is not None:
+            background.paste(foreground, mask=foreground.split()[-1])
+
+    def _generate_cache_filename(self, cache_prefix, area_def, overlays_dict):
+        area_hash = hash(area_def)
+        base_dir, file_prefix = os.path.split(cache_prefix)
+        params_to_hash = self._prepare_hashable_dict(overlays_dict)
+        param_hash = hash_dict(params_to_hash)
+        return os.path.join(base_dir, f"{file_prefix}_{area_hash}_{param_hash}.png")
+
+    @staticmethod
+    def _prepare_hashable_dict(nonhashable_dict):
+        params_to_hash = {}
+        # avoid wasteful deep copy by only doing two levels of copying
+        for overlay_name, overlay_dict in nonhashable_dict.items():
+            if overlay_name == 'cache':
+                continue
+            params_to_hash[overlay_name] = overlay_dict.copy()
+        # font objects are not hashable
+        for font_cat in ('cities', 'points', 'grid'):
+            if font_cat in params_to_hash:
+                params_to_hash[font_cat].pop('font', None)
```

Comment:
I think a warning should be issued here, so that it is clear that the caching ignores possible font changes (and the old font is preserved)

Reply:
Or could the name of the font be used ? can this be fetched from the font object?

Reply:
The problem with a warning is that it would be issued every time someone used a font, whether it was the first time it was being cached, was different than the previous font, or was the same as the previous font.

This method is on the base writer and aggdraw fonts and PIL fonts are different. Until we update aggdraw to be smarter I don't think we can actually grab any information from it in python-space. Aggdraw is still stuck with a new version of libagg that doesn't produce the same results (sitting in the main branch) so I can't release it and don't have time to investigate (don't want to refactor aggdraw code until that gets figured out).

Reply:
I'm quite convinced this will come back to bite us if we don't find a way to make the users aware of the limitation explicitly when it runs. Could we do something like this?
```suggestion
                if 'font' in params_to_hash[font_cat]:
                    logger.info("A change of font in the parameters will not trigger a new rendering of cash.")
                params_to_hash[font_cat].pop('font', None)
```
Otherwise, we could make our own wrapper around the font objects.
Third solution, which might be the easiest to implement: put a visible note about this in the documentation.

Reply:
"visible note": in what way. You mean like `.. note::`? I did add it to the docstring but yeah it could be clearer.

Reply:
yeah I was thinking more like something really standing out. But maybe a note is enough

Reply:
Using the default sphinx theme locally but you get the point:

![image](https://user-images.githubusercontent.com/1828519/129583624-0c4237d9-666b-4f57-8a2f-69420d508709.png)

Reply:
I would move the warning up, before the parameters. Maybe add a new sentence in the general description of the function about the cache, so that the warning will not be too awkward?

Reply:
Good point. Done.

Reply:
Looks good now, thanks